middlewares/retry.go

```diff
@@ -11,24 +11,27 @@ import (
 	"github.com/vulcand/oxy/utils"
 )
 
+// Compile time validation responseRecorder implements http interfaces correctly.
 var (
-	_ http.ResponseWriter = &ResponseRecorder{}
-	_ http.Hijacker       = &ResponseRecorder{}
-	_ http.Flusher        = &ResponseRecorder{}
-	_ http.CloseNotifier  = &ResponseRecorder{}
+	_ http.ResponseWriter = &retryResponseRecorder{}
+	_ http.Hijacker       = &retryResponseRecorder{}
+	_ http.Flusher        = &retryResponseRecorder{}
+	_ http.CloseNotifier  = &retryResponseRecorder{}
 )
 
 // Retry is a middleware that retries requests
 type Retry struct {
 	attempts int
 	next     http.Handler
+	metrics  RetryMetrics
 }
 
 // NewRetry returns a new Retry instance
-func NewRetry(attempts int, next http.Handler) *Retry {
+func NewRetry(attempts int, next http.Handler, metrics RetryMetrics) *Retry {
```

Comment:
Ok, I see your point and I am fine with implementing the abstraction layer. Though I would like to make sure that we are on the same page with our understanding of the code. 

The Retry middleware is not coupled to the metrics middleware. It only knows about the `RetryMetrics` interface, which happens to be right now in the middleware package (IMHO this is the rather confusing part). The `Metrics` and `RetryMetrics` interfaces are in my opinion already a layer of abstraction. The metrics middleware in turn also "only" depends on the `Metrics` interface.

Though the abstraction you suggest would remove all knowledge about metrics from the retry middleware. Do I understand your intention correctly? If so and you suggest that way, I will adapt the PR accordingly.